bindings/rust/src/lazy_reader.rs

```diff
@@ -0,0 +1,518 @@
+        let count = first_offset / 4 - 1;
+        let mut last_offset = None;
+        for _ in 0..count {
+            let offset = cur.unpack_number()?;
+            if last_offset.is_some() && last_offset.unwrap() > offset {
+                return Err(Error::Verify(format!(
+                    "verify_dynvec: invalid offset({} and {})",
+                    last_offset.unwrap(),
+                    offset
+                )));
+            }
+            last_offset = Some(offset);
+            cur.add_offset(NUMBER_SIZE)?;
+        }
+        Ok(())
+    }
+    pub fn verify_fixvec(&self, item_size: usize) -> Result<(), Error> {
+        if self.size < NUMBER_SIZE {
+            return Err(Error::Verify(format!(
+                "verify_fixvec: self.size({}) < NUMBER_SIZE(4)",
+                self.size
+            )));
+        }
+        let item_count = self.unpack_number()?;
+        if item_count == 0 {
+            if self.size == NUMBER_SIZE {
+                return Ok(());
+            } else {
+                return Err(Error::Verify(format!(
+                    "verify_fixvec: self.size({}) != 4",
+                    self.size
+                )));
+            }
+        }
+
+        let total_size = calculate_offset(item_size, item_count, NUMBER_SIZE)?;
+        if self.size == total_size {
+            Ok(())
+        } else {
+            Err(Error::Verify(format!(
+                "verify_fixvec: self.size({}) != total_size({})",
+                self.size, total_size
+            )))
+        }
+    }
+
+    pub fn option_is_none(&self) -> bool {
+        self.size == 0
+    }
+    pub fn fixvec_length(&self) -> Result<usize, Error> {
+        self.unpack_number()
+    }
+
+    pub fn dynvec_length(&self) -> Result<usize, Error> {
+        if self.size == NUMBER_SIZE {
+            Ok(0)
+        } else {
+            let mut cur2 = self.clone();
+            cur2.add_offset(NUMBER_SIZE)?;
+            cur2.sub_size(NUMBER_SIZE)?;
+            cur2.validate()?;
+            cur2.get_item_count()
+        }
+    }
+
+    pub fn get_item_count(&self) -> Result<usize, Error> {
+        let len = self.unpack_number()?;
+        if len % 4 != 0 {
+            return Err(Error::UnknownItem(format!(
+                "get_item_count: not aligned, len = {}",
+                len
+            )));
+        }
+        let count = len / 4;
+        if count == 0 {
+            Err(Error::UnknownItem("get_item_count".into()))
+        } else {
+            Ok(count - 1)
+        }
+    }
+
+    pub fn table_actual_field_count(&self) -> Result<usize, Error> {
+        self.dynvec_length()
+    }
+
+    pub fn table_has_extra_fields(&self, field_count: usize) -> Result<bool, Error> {
+        let count = self.table_actual_field_count()?;
+        Ok(count > field_count)
+    }
+
+    pub fn slice_by_offset(&self, offset: usize, size: usize) -> Result<Cursor, Error> {
+        let mut cur2 = self.clone();
```

Comment:
The code reads better if we simply calculate the new offset, then create a new cursor using new offset, size, and data source.